### PR TITLE
(#3486) - move parseDoc() to own file

### PR DESCRIPTION
**Pull request by nolanlawson**
This is mostly just code cleanup, but also
partially motivated by performance, because
profiling of map/reduce shows that this function
gets called a lot (once per `emit()` basically).

Below I've noted where I made some micro-optimizations.

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -179,103 +122,8 @@ exports.filterChange = function filterChange(opts) {
...
-  var newRevId;
-  var revInfo;
-  var error;
-  var opts = {status: 'available'};
-  if (doc._deleted) {
-    opts.deleted = true;
-  }
-
-  function parseRevisionInfo() {
-    revInfo = /^(\d+)-(.+)$/.exec(doc._rev);
```
Comment (nolanlawson) [inline lib/utils.js]: simplified the regex a bit. no need for the capture groups or `(.+)$` to search all the way to the end of the string

File: lib/utils.js [hunk 2]
```diff
@@ -179,103 +122,8 @@ exports.filterChange = function filterChange(opts) {
...
-  var nRevNum;
-  var newRevId;
-  var revInfo;
-  var error;
-  var opts = {status: 'available'};
-  if (doc._deleted) {
-    opts.deleted = true;
-  }
-
-  function parseRevisionInfo() {
```
Comment (nolanlawson) [inline lib/utils.js]: ^ functions-within-functions are bad for perf. moved this out

File: lib/utils.js [hunk 3]
```diff
@@ -179,103 +122,8 @@ exports.filterChange = function filterChange(opts) {
...
-    if (doc._revisions) {
-      doc._rev_tree = [{
-        pos: doc._revisions.start - doc._revisions.ids.length + 1,
-        ids: doc._revisions.ids.reduce(function (acc, x) {
-          if (acc === null) {
-            return [x, opts, []];
-          } else {
-            return [x, {status: 'missing'}, [acc]];
-          }
-        }, null)
```
Comment (nolanlawson) [inline lib/utils.js]: `reduce()` is bad for perf. replaced with a for-loop.

File: lib/utils.js [hunk 4]
```diff
@@ -179,103 +122,8 @@ exports.filterChange = function filterChange(opts) {
...
-      doc._rev_tree = [{
-        pos: nRevNum,
-        ids: [newRevId, opts, []]
-      }];
-    }
-  }
-
-  exports.invalidIdError(doc._id);
-
-  doc._rev = [nRevNum, newRevId].join('-');
```
Comment (nolanlawson) [inline lib/utils.js]: ^ unnecessary array allocation. replaced with string concatenation

**Comment by nickcolley** [comment]
@nolanlawson do you have the perf stats with this change?

**Comment by nolanlawson** [comment]
Just ran 'em. :)

`temp-views` with 5 iterations:

    Chrome: 77057ms -> 67663ms
    Firefox: 49785ms -> 50868ms

Re-ran the tests just to be safe:

    Chrome: 63215ms -> 62547ms
    Firefox: 52308ms -> 49003ms

I didn't expect a big measurable change, but at least we haven't regressed. I think I would need to do many more iterations to actually figure out how much it's improved. (The first Chrome run was probably a fluke I think.)

**Comment by daleharvey** [comment]
This is cleaner and should be marginally faster, cheers - https://github.com/pouchdb/pouchdb/commit/8d0728682fb73d76e3d2198a1bf9bb30f6b3be35

**Pull request closed without merging**